### PR TITLE
IR visitor for boolean expressions

### PR DESCRIPTION
**Pull request by N-Maas**

### CODE REVIEW

File: src/main/java/edu/kit/compiler/transform/IRBooleanExpressions.java [hunk 1]
```diff
@@ -1,22 +1,272 @@
...
+                throw new IllegalArgumentException();
+            }
+
+            /**
+             * We try to create some optimized code here, because
+             * optimizing conditionals/branches afterwards is not that simple.
+             */
+            ExpressionNode inner = expr.getExpression();
+            if (inner instanceof UnaryExpressionNode) {
+                return evalExpression(context, ((UnaryExpressionNode)inner).getExpression());
```
Comment (g4rgoil) [inline src/main/java/edu/kit/compiler/transform/IRBooleanExpressions.java:76]: You could use `instanceof` pattern matching here. I believe we would need to bump the JDK version to do this though.

File: src/main/java/edu/kit/compiler/transform/IRBooleanExpressions.java [hunk 2]
```diff
@@ -1,22 +1,272 @@
...
+        }
+        return expr.accept(new ValueVisitor(context));
     }
 
     public static void asConditional(TransformContext context, ExpressionNode expr,
                                      Block trueBranch, Block falseBranch) {
-        throw new UnsupportedOperationException();
+        if (!expr.getResultType().equals(boolType)) {
+            throw new IllegalArgumentException();
+        }
```
[the remark this answers is not in the document]
Comment (N-Maas) [reply src/main/java/edu/kit/compiler/transform/IRBooleanExpressions.java:39]: Not really, I can change it

File: src/main/java/edu/kit/compiler/transform/IRBooleanExpressions.java [hunk 3]
```diff
@@ -1,22 +1,272 @@
...
+
+        public Boolean visit(MethodInvocationExpressionNode expr) { return true; }
+        public Boolean visit(FieldAccessExpressionNode expr) { return true; }
+        public Boolean visit(ArrayAccessExpressionNode expr) { return true; }
+        public Boolean visit(IdentifierExpressionNode expr) { return true; }
+        public Boolean visit(ValueExpressionNode expr) { return true; }
+    }
+
+    private static boolean preferAsValue(ExpressionNode expr) {
+        return expr.accept(new ValueOrConditionalDecisionVisitor());
```
[the remark this answers is not in the document]
Comment (N-Maas) [reply src/main/java/edu/kit/compiler/transform/IRBooleanExpressions.java:260]: I could instead construct it at initialization, but there are many cases where this visitor is not even called. So I'm not sure whether that would be better

File: src/main/java/edu/kit/compiler/transform/IRBooleanExpressions.java (hunk 1 above)
Comment (N-Maas) [reply src/main/java/edu/kit/compiler/transform/IRBooleanExpressions.java:76]: And I would need to setup the new JDK version first :sweat_smile:

Comment (g4rgoil) [reply src/main/java/edu/kit/compiler/transform/IRBooleanExpressions.java:73]: Ok, fair enough.

File: src/main/java/edu/kit/compiler/transform/IRBooleanExpressions.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (g4rgoil) [reply src/main/java/edu/kit/compiler/transform/IRBooleanExpressions.java:257]: For anything other than trivial programs I think it's unlikely that the visitor would never be instantiated. So having an instance as a `private static final` attribute might make sense.

[the remark this answers is not in the document]
Comment (N-Maas) [reply src/main/java/edu/kit/compiler/transform/IRBooleanExpressions.java:257]: Thats true

File: src/main/java/edu/kit/compiler/transform/IRBooleanExpressions.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (g4rgoil) [reply src/main/java/edu/kit/compiler/transform/IRBooleanExpressions.java]: There is one more instance in `ValueVisitor#visit(UnaryExpression)` I think you have overlooked.

[the remark this answers is not in the document]
Comment (N-Maas) [reply src/main/java/edu/kit/compiler/transform/IRBooleanExpressions.java]: Seems I have overlooked even two more instances^^

**Review by g4rgoil — APPROVED** [review]
Looks good to me. Obviously the placeholder code needs to be removed once possible.

**Pull request merged**